### ISSUE TITLE
Question about QoS levels for State topic.

### ISSUE THREAD

**Issue by gauspohl** [VDA5050/VDA5050#237]
I see that the Connection topic was originally QoS level zero, and was changed to 1 per the following issue: https://github.com/VDA5050/VDA5050/issues/24

We are wondering why we don't utilize QoS level one for the State topic as well? As is, state information can conceivably be lost, such as if a vehicle is taken from automatic into manual and thereafter back into automatic. If we would like to have some functionality in the traffic controller take place when a vehicle is placed into manual mode, we'd have to build an abstraction on top of the protocol to ensure this which is inelegant.

**Comment by LukasSchwarz4202** [comment]
QoS level 1 has more communication overhead compared to QoS level 0. Still tests showed us, that the overhead is minimal and we fully support the QoS for the state topic as well. Actually we would suggest to implement it for every topic.

**Comment by gauspohl** [comment]
Update from WG01 meeting on 2023/11/08: 

Changing to QoS 1 only ensures broker -> MC connection. It would still not be guaranteed the message makes it from the truck to the broker. It is already the case in VDA5050 that state messages can be swallowed if the states change very rapidly.

The implicit intention of VDA5050 is that if you want the full historical state data of a vehicle, you must ask the vehicle itself for this outside of the scope of VDA5050. 

I will take this info back to my team and then update/close this issue as necessary.

**Comment by jiaweirobot** [comment]
Hello, everyone.
I also have some questions about this Qos leave
Does the above mean that I need to have a method to query the vehicle at regular intervals and keep sending nstant action to the vehicle to query the status?

**Comment by collani-bosch** [comment]
Some thoughts to start the discussion again:

- The QoS 1 ensures that a MQTT message is received at least once by the addressee. This is done by resending the message if the sender doesn’t receive an acknowledge in time. Because the sender ensures the order of the messages this might lead to sending more than one message again. E.g. Broker sends message “A”, “B”, “C”, “D” and gets no acknowledge for “B” the broker will send “B”, “C” and “D” again (in this order). The addressee will get message “C” and “D” twice (or more).
- Typically the connection to a broker is handled by a TCP/IP connection and the TCP/IP Stack ensures that the application (the MQTT client or broker) sees a correct data stream. Handling of lost or out-of-order Ethernet packages are handled by the TCP/IP stack. The only way to lose MQTT messages is a broken TCP/IP connection. Typically this will happen when the MQTT client or broker received no "heartbeat" messages and closes the TCP/IP connection (the TCP/IP stack itself of very patient before giving up a connection).
- To receive outstanding messages of a QoS 1 connection the client has to use single MQTT session id. Otherwise the broker cannot map the outstanding message to the reconnected client. Deserted sessions also stresses the broker, because for all sessions with QoS 1 (or higher) messages have to be buffered (and MQTT V3.1.1 defines no end-of-life for a session).

Using QoS 1 can be tricky (I speak from experience.).

- Using QoS 1 for the connection topic makes sense, because messages on these topics will only send once (these messages have an event nature).
- The state message contains the complete state of the vehicle (not a singe event) and is send at least ever 20 - 30 sec (with zones more often) and can activity requested by MC. No harm is done if some message is lost. Maybe it should be best practices to send a state message immediately after a reconnect to the broker.
- The visualization message: "Master control must not use the info messages for logic, it must only be used for visualization and debugging purposes." and why (re-)sending outdated messages.
- Using QoS 1 for the instant actions and order messages may useful to unburden MC with a logic to re-transmit an instant action or order message (the MQTT protocol will do it). (From my observations, it seems to be difficult enough to interpret the status message correctly.) In this case the handling of the 'headerId' has to be defined precisely, because messages can be received twice or more. Make it sense to introduce this now for these topics, because today MC should have a mechanism to send instant actions and order messages again?

**Comment by camil22** [comment]
I dont think that introducing QoS 1 for instant actions and orders would remove the need to resend said messages. QoS 1 would only ensure that these messsages would reach the broker but when the AMR is disconnected at that time the message would never reach it. (I assume that at the time of sending the messages MC isnt aware of the AMR being disconnected)

In that case MC still needs to resend these messages after the AMR reconnects and does not reflect them in its state.

**Comment by collani-bosch** [comment]
QoS 1 is only effective if each client use it; especially the client connected via WiFi to the broker. If this is the case and the client (in our use case the AGV) loses the connection to the broker, the broker buffers each message for the client and send the message again until the broker receives an acknowledge form the client. Therefore it is crucial that the client sticks to his session id thereby the broker can detect the reconnection and resend the message.
Because of the necessary buffering QoS1 leads to a higher resource consumption (especially at the broker) and if the client uses a new session id for every connect it can be really messy on broker side, because the broker has to buffer messages for open sessions (MQTT V3 doesn't specify the lifetime of a session).

**Comment by ConstantinUp** [comment]
This topic was not touched for version 3.0. The document is currently design to work fine with a QoS 0. We might circle back to this discussions in future versions, e.g., in combination with message serialization.

If further points arise, please start a new issue or reopen this one.
